hw3/02-charts.js

```diff
@@ -1,52 +1,105 @@
 const backgroundColors = [
-  'rgba(54, 162, 235, 0.8)',
-  'rgba(255, 206, 86, 0.8)',
-  'rgba(255, 99, 132, 0.8)',
-  'rgba(75, 192, 192, 0.8)',
-  'rgba(153, 102, 255, 0.8)',
-  'rgba(255, 159, 64, 0.8)',
-  'rgba(199, 199, 199, 0.8)',
-  'rgba(83, 102, 255, 0.8)',
-  'rgba(40, 159, 64, 0.8)',
-  'rgba(210, 199, 199, 0.8)',
-  'rgba(78, 52, 199, 0.8)',
+	"rgba(54, 162, 235, 0.8)",
+	"rgba(255, 206, 86, 0.8)",
+	"rgba(255, 99, 132, 0.8)",
+	"rgba(75, 192, 192, 0.8)",
+	"rgba(153, 102, 255, 0.8)",
+	"rgba(255, 159, 64, 0.8)",
+	"rgba(199, 199, 199, 0.8)",
+	"rgba(83, 102, 255, 0.8)",
+	"rgba(40, 159, 64, 0.8)",
+	"rgba(210, 199, 199, 0.8)",
+	"rgba(78, 52, 199, 0.8)",
 ];
 
 const borderColors = [
-  'rgba(54, 162, 235, 1)',
-  'rgba(255, 206, 86, 1)',
-  'rgba(255, 99, 132, 1)',
-  'rgba(75, 192, 192, 1)',
-  'rgba(153, 102, 255, 1)',
-  'rgba(255, 159, 64, 1)',
-  'rgba(159, 159, 159, 1)',
-  'rgba(83, 102, 255, 1)',
-  'rgba(40, 159, 64, 1)',
-  'rgba(210, 199, 199, 1)',
-  'rgba(78, 52, 199, 1)',
+	"rgba(54, 162, 235, 1)",
+	"rgba(255, 206, 86, 1)",
+	"rgba(255, 99, 132, 1)",
+	"rgba(75, 192, 192, 1)",
+	"rgba(153, 102, 255, 1)",
+	"rgba(255, 159, 64, 1)",
+	"rgba(159, 159, 159, 1)",
+	"rgba(83, 102, 255, 1)",
+	"rgba(40, 159, 64, 1)",
+	"rgba(210, 199, 199, 1)",
+	"rgba(78, 52, 199, 1)",
 ];
 
 // url for the Thrones API
-const url = 'https://thronesapi.com/api/v2/Characters';
-
-const renderChart = () => {
-  const donutChart = document.querySelector('.donut-chart');
-
-  new Chart(donutChart, {
-    type: 'doughnut',
-    data: {
-      labels: ['label', 'label', 'label', 'label'],
-      datasets: [
-        {
-          label: 'My First Dataset',
-          data: [1, 12, 33, 5],
-          backgroundColor: backgroundColors,
-          borderColor: borderColors,
-          borderWidth: 1,
-        },
-      ],
-    },
-  });
+const url = "https://thronesapi.com/api/v2/Characters";
+
+//fetch api
+const fetchData = async (url) => {
+	try {
+		const response = await fetch(url);
+		const apiData = await response.json();
+
+		const allFamilies = apiData.map((character) => character.family);
+
+		allFamilies.forEach((house, index) => {
+			if (house.includes("Lanister")) {
+				allFamilies[index] = "House Lannister";
+			}
+		});
+
+		const houseCounts = allFamilies.reduce((accumulator, house) => {
+			let existingHouse = false;
+
+			for (let key in accumulator) {
+				if (key.includes(house)) {
+					accumulator[key] += 1;
+					existingHouse = true;
+					break;
+				}
+				if (!key.startsWith("House") && accumulator[key] == 1) {
+					accumulator["Others"] = 1;
+					existingHouse = true;
+					break;
+				}
+				if (key == "None" || key == "Unknown" || key == "Unkown") {
+					accumulator["Unknown"] += 1;
+					existingHouse = true;
+					break;
+				}
+			}
+			if (!existingHouse) {
+				accumulator[house] = 1;
+			}
+
+			return accumulator;
+		}, {});
```

Comment:
I think it is funny how everyone approached this problem in a different way, I used regex for part of mine and almost everyone I have seen has used different conditional statements for clearing out bad data. Looks like yours works though, I honestly don't think there is a clean way to do this

Reply:
Haha, Yeah. Had a hard time trying to figure out how to preprocess the API data. Implementation through regex sounds very efficient though!